File: algorithms/81.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    bool search(vector<int>& nums, int target)
    {
        int l = 0, r = nums.size()-1;
        while ( l <= r )
        {
            while ( l < r && nums[l] == nums[l+1] ) l += 1;
            while ( l < r && nums[r] == nums[r-1] ) r -= 1;
            int m = (l+r)/2;
            if ( nums[m] == target ) return true;
            else if ( target < nums[m] )
            {
                if ( nums[l] <= target || nums[m] < nums[l] )
                    r = m-1;
                else
                    l = m+1;
            }
            else
            {
                if ( nums[l] > target || nums[m] >= nums[l] )
                    l = m+1;
                else
                    r = m-1;
            }
        }
        return false;
    }
};

int main()
{
    return 0;
}

File: algorithms/rust/81.rs
impl Solution {
    #[allow(dead_code)]
    pub fn search(nums: Vec<i32>, target: i32) -> bool {
        if nums[0] == target {
            return true;
        }
        let (mut left, mut right) = (0, nums.len() as i32 - 1);
        while right >= left && nums[left as usize] == nums[right as usize] {
            right -= 1;
        }
        while left <= right {
            let mid = left + (right - left) / 2;
            let mid_value = nums[mid as usize];
            if mid_value == target {
                return true;
            }
            if mid_value >= nums[0] && (target > mid_value || target < nums[0])
                || target < nums[0] && target > mid_value
            {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        false
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
